Task 2 Integrating WebVitals

kt-12 opened the pull request:
## PR Checklist
<!-- 
Please check if your PR fulfills the following requirements:
**Filling out the template is required.** Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
 -->
- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md


## PR Type
What kind of change does this PR introduce?
<!-- Please check the one that applies to this PR using "x" with no spaces eg: [x]. -->
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

kt-12 committed:
Task 2 WebVitals

kt-12 committed:
enqueue script and merge base branch

kt-12 committed:
attempt to enqueue webpack

kt-12 committed:
lcp calculation logic

kt-12 committed:
temp web-package import

kt-12 commented on .grunt-config/webpack.js:
```diff
@@ -139,6 +141,7 @@ const externals = {
 	'@wordpress/i18n': 'wp.i18n',
 	react: 'React',
 	'react-dom': 'ReactDOM',
+	'web-vitals': 'WebVitals',
```
TODO: `get import web-vitials working from node modules` 

cc: @joemcgill

joemcgill replied:
This should be working as of 3741e97fd

kt-12 commented on .grunt-config/webpack.js:
```diff
@@ -152,6 +155,7 @@ const plugins = [
 	new webpack.ProvidePlugin( {
 		React: 'react',
 		ReactDOM: 'react-dom',
+		WebVitals: 'web-vitals',
```
TODO: `get import web-vitials working from node modules`

cc: @joemcgill

joemcgill replied:
See: https://github.com/kt-12/elementor/pull/3#discussion_r1326411859

kt-12 commented on modules/image-loading-optimization/assets/js/editor/hooks/check-lcp-image-before-save.js:
```diff
@@ -0,0 +1,138 @@
+        currentScrollPosition = this.getPreviewFrame().scrollY;
+
+        // Switch to desktop mode to get the LCP image.
+        const deviceMode = 'desktop'
```
@joemcgill  Currently only supports only desktop. I did try switching to other device mode but it was not capturing new LCP even after page reload inside iframe using `elementor.reloadPreview();` possible issue with `DOMContentLoaded` only called once ( [this line ](https://github.com/kt-12/elementor/pull/3/files#diff-85e71fb51b8e4fbbd1b086e2e213f371f627ab0b9cdf486af9b5f2f3495282f8R85) )

kt-12 commented on modules/image-loading-optimization/assets/js/editor/hooks/check-lcp-image-before-save.js:
```diff
@@ -0,0 +1,138 @@
+        return false;
+    }
+
+    isElementAboveTheFold( element ) {
```
This is currently above the fold for desktop mode only. If we make device specific we need to figure that out.

kt-12 commented on modules/image-loading-optimization/assets/js/editor/hooks/check-lcp-image-before-save.js:
```diff
@@ -0,0 +1,138 @@
+	 *
+	 * @return {Window}
+	 */
+	getPreviewFrame() {
```
method borrowed from 

`elementor/modules/styleguide/assets/js/e-component.js`

kt-12 commented on modules/image-loading-optimization/assets/js/preview/index.js:
```diff
@@ -0,0 +1,85 @@
+import { onLCP } from './web-vitals';
```
Currently, `web-vitals.js` is placed in same folder. Couldn't figure out why direct import is not working.

cc: @joemcgill

kt-12 commented on modules/image-loading-optimization/assets/js/preview/index.js:
```diff
@@ -0,0 +1,85 @@
+import { onLCP } from './web-vitals';
+
+class WebVitalsChecker extends elementorModules.ViewModule {
```
extending `elementorModules.ViewModule` not mandatory here, as we are only using `onInit` from `elementorModules.ViewModule`

joemcgill committed:
Import web-vitals library as a dependency

This updates the webpack config so the web vitals library is not treated like an external during bundling so the onLCP function can be imported directly.

joemcgill committed:
Automatic formatting fixes

joemcgill committed:
Export and name JS module

joemcgill committed:
Improve JS module loading

This uses the new loading strategies added in a backward compatible way in WP 6.3 when registering editor scripts and avoids bundling global resources in the module script for some added performance benefits.

joemcgill commented on modules/image-loading-optimization/assets/js/editor/component.js:
```diff
@@ -1,8 +1,7 @@
 
-import ComponentBase from 'elementor-api/modules/component-base';
```
I noticed that importing this made the bundled module file pretty large, while extending it using the global this way made the file around 20kb smaller. This pattern happens in a lot of places in Elementor, so adding some code splitting could be a nice benefit, but outside the scope of this PR for now.

kt-12 commented on modules/image-loading-optimization/module.php:
```diff
@@ -336,4 +342,38 @@ private function high_priority_element_flag( $value = null ) {
+			$this->get_js_assets_url( $this->get_name() . '-preview' ),
+			[],
+			ELEMENTOR_VERSION,
+			array(
```
Since elementor would also support older version of WordPress we need to have a version check for `>=6.3` here and only then have the strategy.
